File: dispatch/java/runtime/adamma/c4dhi/claid/Module/Scheduling/ScheduleHelper.java
package adamma.c4dhi.claid.Module.Scheduling;

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

import adamma.c4dhi.claid.Schedule;
import adamma.c4dhi.claid.SchedulePeriodic;
import adamma.c4dhi.claid.ScheduleTimeOfDay;

public class ScheduleHelper {
    public static Duration getIntervalDurationFromPeriodicSchedule(SchedulePeriodic schedulePeriodic) {
        switch (schedulePeriodic.getIntervalCase()) {
            case FREQUENCY_HZ:
                return Duration.ofMillis((long) (1000.0 / schedulePeriodic.getFrequencyHz()));
            case FREQUENCY_KHZ:
                return Duration.ofMillis((long) (1000.0 / (schedulePeriodic.getFrequencyKHz() * 1000.0)));
            case FREQUENCY_MHZ:
                return Duration.ofMillis((long) (1000.0 / (schedulePeriodic.getFrequencyMHz() * 1000000.0)));
            case PERIOD_MILLISECONDS:
                return Duration.ofMillis((long) schedulePeriodic.getPeriodMilliseconds());
            case PERIOD_SECONDS:
                return Duration.ofSeconds((long) schedulePeriodic.getPeriodSeconds());
            case PERIOD_MINUTES:
                return Duration.ofMinutes((long) schedulePeriodic.getPeriodMinutes());
            case PERIOD_HOURS:
                return Duration.ofHours((long) schedulePeriodic.getPeriodHours());
            case PERIOD_DAYS:
                return Duration.ofDays((long) schedulePeriodic.getPeriodDays());
            default:
                return Duration.ZERO;
        }
    }

    public static LocalDateTime calculateNextTimeOfDay(ScheduleTimeOfDay timeOfDay) {
        // Get the current date and time
        LocalDateTime now = LocalDateTime.now();

        // Extract current date and time components
        LocalDate today = now.toLocalDate();

        // Extract target time components from ScheduleTimeOfDay
        int targetHour = timeOfDay.getHour();
        int targetMinute = timeOfDay.getMinute();
        int targetSecond = timeOfDay.getSecond();

        // Create a LocalTime object for the target time of day
        LocalTime targetTime = LocalTime.of(targetHour, targetMinute, targetSecond);

        // Create a LocalDateTime object for today at the target time
        LocalDateTime targetDateTime = LocalDateTime.of(today, targetTime);

        // If the target time has already passed for today, schedule it for tomorrow
        if (now.isAfter(targetDateTime)) {
            targetDateTime = targetDateTime.plusDays(1);
        }

        return targetDateTime;
    }

    public static boolean isFirstTimeOfDayBeforeSecond(ScheduleTimeOfDay first, ScheduleTimeOfDay second) {
        LocalTime firstTime = LocalTime.of(first.getHour(), first.getMinute(), first.getSecond());
        LocalTime secondTime = LocalTime.of(second.getHour(), second.getMinute(), second.getSecond());
        return firstTime.isBefore(secondTime);
    }

    public static boolean areTimesOfDayEqual(ScheduleTimeOfDay first, ScheduleTimeOfDay second) {
        return first.getHour() == second.getHour() &&
                first.getMinute() == second.getMinute() &&
                first.getSecond() == second.getSecond();
    }

    public static boolean isScheduleEmpty(Schedule schedule)
    {
        return schedule.getPeriodicList().size() == 0 && schedule.getTimedList() == null;
    }

}

File: dispatch/ios/CLAID/Sources/CLAID/Module/Scheduling/ScheduleHelper.swift
//
//  ScheduleHelper.swift
//  CLAID
//
//  Created by Patrick Langer on 19.02.2025.
//

import Foundation

public class ScheduleHelper {
    

    public static func getIntervalDurationFromPeriodicSchedule(_ schedulePeriodic: Claidservice_SchedulePeriodic) -> Duration {
        switch schedulePeriodic.interval {
        case .frequencyHz:
            return Duration.milliseconds(Int64(1000.0 / schedulePeriodic.frequencyHz))
        case .frequencyKHz:
            return Duration.milliseconds(Int64(1000.0 / (schedulePeriodic.frequencyKHz * 1000.0)))
        case .frequencyMhz:
            return Duration.milliseconds(Int64(1000.0 / (schedulePeriodic.frequencyMhz * 1_000_000.0)))
        case .periodMilliseconds:
            return Duration.milliseconds(Int64(schedulePeriodic.periodMilliseconds))
        case .periodSeconds:
            return Duration.seconds(Int64(schedulePeriodic.periodSeconds))
        case .periodMinutes:
            return Duration.seconds(Int64(schedulePeriodic.periodMinutes * 60))
        case .periodHours:
            return Duration.seconds(Int64(schedulePeriodic.periodHours * 60 * 60))
        case .periodDays:
            return Duration.seconds(Int64(schedulePeriodic.periodDays * 60 * 60 * 24)) // Duration doesn't have a "days" unit, so convert to hours
        default:
            return .zero
        }
    }

    public static func calculateNextTimeOfDay(_ timeOfDay: Claidservice_ScheduleTimeOfDay) -> Date {
        let now = Date()
        let calendar = Calendar.current
        
        var dateComponents = calendar.dateComponents([.year, .month, .day], from: now)
        dateComponents.hour = Int(Int32(timeOfDay.hour))
        dateComponents.minute = Int(Int32(timeOfDay.minute))
        dateComponents.second = Int(Int32(timeOfDay.second))
        
        if let targetDateTime = calendar.date(from: dateComponents), targetDateTime > now {
            return targetDateTime
        }
        
        return calendar.date(byAdding: .day, value: 1, to: dateComponents.date!)!
    }

    public static func isFirstTimeOfDayBeforeSecond(_ first: Claidservice_ScheduleTimeOfDay, _ second: Claidservice_ScheduleTimeOfDay) throws -> Bool {
        guard let firstTime = Calendar.current.date(from: DateComponents(
            hour: Int(first.hour),
            minute: Int(first.minute),
            second: Int(first.second)
        )), let secondTime = Calendar.current.date(from: DateComponents(
            hour: Int(second.hour),
            minute: Int(second.minute),
            second: Int(second.second)
        )) else {
            throw CLAIDError("Invalid dates specified in config for schedule: \(first), \(second)")
        }
        return firstTime < secondTime
    }

    public static func areTimesOfDayEqual(_ first: Claidservice_ScheduleTimeOfDay, _ second: Claidservice_ScheduleTimeOfDay) -> Bool {
        return first.hour == second.hour &&
               first.minute == second.minute &&
               first.second == second.second
    }


    public static func isScheduleEmpty(_ schedule: Claidservice_Schedule) -> Bool {
        return schedule.periodic.isEmpty && schedule.timed.isEmpty
    }
}
